File: animal.py
"""Animal/Dog/Cat/Bird demo — mirrors animal.go, for exploring aerial.nvim's
outline (same shapes: an interface-like protocol, three concrete types, free
functions, a composed Zoo, a generic-ish helper, and a nested function inside
main)."""

from typing import Protocol


class Animal(Protocol):
    def sound(self) -> str: ...
    def get_name(self) -> str: ...


class Dog:
    def __init__(self, name: str) -> None:
        self.name = name

    def sound(self) -> str:
        return "woof"

    def get_name(self) -> str:
        return self.name


class Cat:
    def __init__(self, name: str) -> None:
        self.name = name

    def sound(self) -> str:
        return "meow"

    def get_name(self) -> str:
        return self.name


class Bird:
    def __init__(self, name: str) -> None:
        self.name = name

    def sound(self) -> str:
        return "tweet"

    def get_name(self) -> str:
        return self.name

    def fly(self) -> str:
        return f"{self.name} flies away"


def describe(animal: Animal) -> str:
    return f"{animal.get_name()} says {animal.sound()}"


class Zoo:
    """Composition: holds a list of Animals, like Go's Zoo struct."""

    def __init__(self, animals: list[Animal]) -> None:
        self.animals = animals

    def describe(self) -> str:
        out = ""
        for animal in self.animals:
            out += describe(animal) + "\n"
        return out


def max_value(a, b):
    """Python has no type-parameter syntax like Go's Max[T] — duck typing covers it."""
    return a if a > b else b


def main() -> None:
    dog = Dog("Rex")
    cat = Cat("Whiskers")
    bird = Bird("Tweety")

    dog.sound()
    cat.sound()
    bird.fly()

    describe(dog)
    describe(cat)
    describe(bird)

    def greet(animal: Animal) -> str:
        """Nested function — unlike Go's anonymous closure, this IS a named
        `def` inside another `def`; worth checking whether aerial nests it."""
        return f"Hello, {animal.get_name()}!"

    greet(dog)

    zoo = Zoo([dog, cat, bird])
    zoo.describe()

    max_value(3, 7)


if __name__ == "__main__":
    main()

File: animal.rs
//! Animal/Dog/Cat/Bird demo — mirrors animal.go, for exploring aerial.nvim's
//! outline. Rust methods live inside an `impl` block, which is real lexical
//! containment (unlike Go's methods, which sit outside the struct entirely) —
//! this is the interesting middle ground: does aerial nest a method under its
//! `impl Dog { ... }` block? (Bird has two impl blocks — inherent + trait —
//! to see how aerial handles both under one struct.)

trait Animal {
    fn sound(&self) -> String;
    fn name(&self) -> String;
}

struct Dog {
    name: String,
}

impl Animal for Dog {
    fn sound(&self) -> String {
        "woof".to_string()
    }

    fn name(&self) -> String {
        self.name.clone()
    }
}

struct Cat {
    name: String,
}

impl Animal for Cat {
    fn sound(&self) -> String {
        "meow".to_string()
    }

    fn name(&self) -> String {
        self.name.clone()
    }
}

struct Bird {
    name: String,
}

impl Bird {
    fn fly(&self) -> String {
        format!("{} flies away", self.name)
    }
}

impl Animal for Bird {
    fn sound(&self) -> String {
        "tweet".to_string()
    }

    fn name(&self) -> String {
        self.name.clone()
    }
}

fn describe(animal: &dyn Animal) -> String {
    format!("{} says {}", animal.name(), animal.sound())
}

/// Composition: holds a Vec of boxed Animals, like Go's Zoo struct.
struct Zoo {
    animals: Vec<Box<dyn Animal>>,
}

impl Zoo {
    fn describe(&self) -> String {
        let mut out = String::new();
        for animal in &self.animals {
            out += &describe(animal.as_ref());
            out += "\n";
        }
        out
    }
}

/// Generic function — Rust's real type-parameter syntax.
fn max_value<T: PartialOrd>(a: T, b: T) -> T {
    if a > b {
        a
    } else {
        b
    }
}

fn main() {
    let dog = Dog { name: "Rex".to_string() };
    let cat = Cat { name: "Whiskers".to_string() };
    let bird = Bird { name: "Tweety".to_string() };

    dog.sound();
    cat.sound();
    bird.fly();

    describe(&dog);
    describe(&cat);
    describe(&bird);

    // Closure assigned to a local binding — like Go's anonymous func in main.
    let greet = |animal: &dyn Animal| format!("Hello, {}!", animal.name());
    greet(&dog);

    let zoo = Zoo {
        animals: vec![Box::new(dog), Box::new(cat), Box::new(bird)],
    };
    zoo.describe();

    max_value(3, 7);
}
